torch/csrc/autograd/record_function.cpp

```diff
@@ -0,0 +1,83 @@
+#include <torch/csrc/autograd/record_function.h>
+#include <torch/csrc/autograd/function.h>
+
+namespace torch { namespace autograd { namespace profiler {
+
+namespace {
+bool has_callbacks = false;
+std::vector<RecordFunctionCallback> start_callbacks;
+std::vector<RecordFunctionCallback> end_callbacks;
+thread_local RecordFunction* thread_local_func_ = nullptr;
+}
+
+void pushCallback(RecordFunctionCallback start, RecordFunctionCallback end) {
+  start_callbacks.push_back(start);
+  end_callbacks.push_back(end);
+  has_callbacks = true;
+}
+
+void pushCallback(RecordFunctionCallback start) {
+  pushCallback(start, [](const RecordFunction&){});
+}
+
+void popCallback() {
+  if (start_callbacks.empty()) {
+    throw std::runtime_error("Empty callbacks stack");
+  }
+  start_callbacks.pop_back();
+  end_callbacks.pop_back();
+  has_callbacks = !start_callbacks.empty();
+}
+
+RecordFunction::RecordFunction(Function* fn, GetPackedInputsCallback cb) {
+  if (!has_callbacks) {
+    return;
+  }
+  fn_ = fn;
+  name_ = StringView(fn->name());
+  sequence_nr_ = fn->sequence_nr();
+  inputs_cb_ = cb;
+  processCallbacks();
+}
+
+RecordFunction::RecordFunction(
+    std::string name, int64_t sequence_nr, GetPackedInputsCallback cb) {
+  if (!has_callbacks) {
+    return;
+  }
+  name_ = StringView(std::move(name));
+  sequence_nr_ = sequence_nr;
+  inputs_cb_ = cb;
+  processCallbacks();
+}
+
+RecordFunction::RecordFunction(
+    const char* name, int64_t sequence_nr, GetPackedInputsCallback cb) {
+  if (!has_callbacks) {
+    return;
+  }
+  name_ = StringView(name);
+  sequence_nr_ = sequence_nr;
+  inputs_cb_ = cb;
+  processCallbacks();
+}
+
+void RecordFunction::processCallbacks() {
+  parent_ = thread_local_func_;
+  thread_local_func_ = this;
```

Comment:
Why do we even maintain this? It doesn't seem to be used anywhere.

Reply:
this allows you to check what nested RecordFunction you are in, I've added an example in the test how it is used

Reply:
Can you point me to the exact lines? When I search for this identifier in this patch then the only usages I can see are inside this file.

Reply:
please check the `getFullName` function in `test_misc.h`